File: apps/api/src/common/phone.ts
import { BadRequestException } from '@nestjs/common';
import { parsePhoneNumberWithError, type CountryCode } from 'libphonenumber-js';
/**
 * Clé de déduplication de Representant et Prospect, adossée à un index unique
 * PARTIEL en base : elle doit rester déterministe. L'app mobile applique la même
 * normalisation avec la même région par défaut ; un écart rendrait la
 * déduplication aléatoire.
 */

const DEFAULT_REGION_FALLBACK = 'SN';

const SEPARATORS = /[\s.\-() ‐-―/]/g;

const phoneDefaultRegion = (): CountryCode =>
  (process.env.PHONE_DEFAULT_REGION ?? DEFAULT_REGION_FALLBACK).toUpperCase() as CountryCode;

/**
 * libphonenumber ne reconnaît ni `00221…` ni `221…` avec la région SN par défaut :
 * il les lit comme des numéros locaux absurdes, d'où une E.164 distincte du même
 * abonné.
 */
function canonicalizePrefix(raw: string, region: CountryCode): string {
  const compact = raw.replace(SEPARATORS, '');
  if (compact.startsWith('+')) return compact;

  if (compact.startsWith('00')) return `+${compact.slice(2)}`;

  const callingCode = REGION_CALLING_CODES[region];
  if (callingCode && compact.startsWith(callingCode)) {
    // Longueur vérifiée : sinon un numéro national commençant par les mêmes
    // chiffres que l'indicatif serait amputé.
    const national = compact.slice(callingCode.length);
    if (national.length >= 8) return `+${compact}`;
  }

  return compact;
}

const REGION_CALLING_CODES: Readonly<Record<string, string>> = {
  SN: '221',
  ML: '223',
  CI: '225',
  GN: '224',
  MR: '222',
  GM: '220',
  BF: '226',
  FR: '33',
};

export function normalizePhone(input: string, region: CountryCode = phoneDefaultRegion()): string {
  if (typeof input !== 'string' || input.trim() === '') {
    throw new BadRequestException({
      code: 'PHONE_INVALID',
      message: 'Le numéro de téléphone est obligatoire.',
    });
  }

  const candidate = canonicalizePrefix(input, region);

  try {
    const parsed = parsePhoneNumberWithError(candidate, region);
    if (!parsed.isValid()) {
      throw new BadRequestException({
        code: 'PHONE_INVALID',
        message: `Numéro de téléphone invalide : ${input}`,
      });
    }
    return parsed.number;
  } catch (error) {
    if (error instanceof BadRequestException) throw error;
    throw new BadRequestException({
      code: 'PHONE_INVALID',
      message: `Numéro de téléphone invalide : ${input}`,
    });
  }
}

export function tryNormalizePhone(
  input: string | undefined | null,
  region: CountryCode = phoneDefaultRegion(),
): string | undefined {
  if (!input) return undefined;
  try {
    return normalizePhone(input, region);
  } catch {
    return undefined;
  }
}

File: apps/go/phone.go
package main

import (
	"net/http"
	"regexp"
	"strings"

	"github.com/nyaruka/phonenumbers"
)

// Clé de déduplication de Representant et Prospect, adossée à un index unique
// partiel : même canonisation que common/phone.ts, sinon la base existante
// devient incohérente.
var (
	separateursTelephone = regexp.MustCompile(`[\s.\-()‐‑‒–—―/]`)
	indicatifsRegion     = map[string]string{"SN": "221", "ML": "223", "CI": "225", "GN": "224", "MR": "222", "GM": "220", "BF": "226", "FR": "33"}
)

func canoniserPrefixe(brut, region string) string {
	compact := separateursTelephone.ReplaceAllString(brut, "")
	if strings.HasPrefix(compact, "+") {
		return compact
	}
	if strings.HasPrefix(compact, "00") {
		return "+" + compact[2:]
	}
	if indicatif := indicatifsRegion[region]; indicatif != "" && strings.HasPrefix(compact, indicatif) && len(compact)-len(indicatif) >= 8 {
		return "+" + compact
	}
	return compact
}

func telephoneInvalide(brut string) *ProblemError {
	if strings.TrimSpace(brut) == "" {
		return problem(http.StatusBadRequest, "PHONE_INVALID", "Le numéro de téléphone est obligatoire.")
	}
	return problem(http.StatusBadRequest, "PHONE_INVALID", "Numéro de téléphone invalide : "+brut)
}

func normaliserTelephone(brut, region string) (string, error) {
	if strings.TrimSpace(brut) == "" {
		return "", telephoneInvalide(brut)
	}
	num, err := phonenumbers.Parse(canoniserPrefixe(brut, region), region)
	if err != nil || !phonenumbers.IsValidNumber(num) {
		return "", telephoneInvalide(brut)
	}
	return phonenumbers.Format(num, phonenumbers.E164), nil
}

func telephoneOptionnel(brut *string, region string) *string {
	if brut == nil {
		return nil
	}
	e164, err := normaliserTelephone(*brut, region)
	if err != nil {
		return nil
	}
	return &e164
}
